src/AutoSaveStatus/AutoSaveStatusProvider.tsx

```diff
@@ -0,0 +1,84 @@
+import React, { PropsWithChildren, useCallback, useEffect, useRef, useState } from "react";
+
+export enum AutoSaveStatus {
+  IDLE = "idle",
+  SAVING = "saving",
+  DONE = "done",
+  ERROR = "error",
+}
+
+export interface AutoSaveStatusContextType {
+  status: AutoSaveStatus;
+  /** Resets status to IDLE, particularly useful if "Error" or "Done" is stale */
+  resetStatus: VoidFunction;
+  errors: unknown[];
+  /** Notifies AutoSaveContext that a request is in-flight */
+  triggerAutoSave: VoidFunction;
+  /** Notifies AutoSaveContext that a request has settled, optionally taking an error */
+  resolveAutoSave: (error?: unknown) => void;
+}
+
+export const AutoSaveStatusContext = React.createContext<AutoSaveStatusContextType>({
+  status: AutoSaveStatus.IDLE,
+  resetStatus() {},
+  errors: [],
+  triggerAutoSave() {},
+  resolveAutoSave() {},
+});
+
+type AutoSaveStatusProviderProps = PropsWithChildren<{
+  /** After a successful save, reset Status back to `Idle` after this many milliseconds */
+  resetToIdleTimeout?: number;
+}>;
+
+export function AutoSaveStatusProvider({ children, resetToIdleTimeout = 6_000 }: AutoSaveStatusProviderProps) {
+  const [status, setStatus] = useState(AutoSaveStatus.IDLE);
+  const [errors, setErrors] = useState<unknown[]>([]);
+  const [inFlight, setInFlight] = useState(0);
+  const resetToIdleTimeoutRef = useRef<number | null>(null);
+
+  /** Handles setting Status */
+  useEffect(() => {
+    if (inFlight > 0) return setStatus(AutoSaveStatus.SAVING);
+    if (status === AutoSaveStatus.IDLE) return;
+    if (errors.length) return setStatus(AutoSaveStatus.ERROR);
```

Comment:
I kinda wonder if this should win, but also I guess the assumption is that the "next/in-flight auto save" will fix it?

Makes me wonder about having two separate operations on the page, like I'm auto-saving row 1 and row 2; row 1 auto-saves and fails, row 2 auto-saves and works, should row 2 auto-saving throw away / hide the fact that row 1 auto-save failed?

We do already have tables with per-row auto-saves, but I guess we probably haven't seen what they want the UX for auto-saves for that to be @bdow ? Like will there be a per-row/per-cell auto-save indicator, or a global one up top?

If it's global, I wonder if `triggerAutoSave` needs to take some sort of key, so we could differentiate each in-flight operation; i.e. our internal data structures would move to some sort of map of `operationKey --> inFlight: yes/no, error: set/unset`.

Not for this PR, but just thinking of what that should look like.